# Project 2: Data Import - Working with Web APIs and JSON (Movies Dataset)

## Import modules and packages

In [37]:
import pandas as pd
import json
import requests
from functools import partial, reduce

## Importing Data from JSON files 

1. __Import__ the json files __blockbusters.json__, __blockbusters2.json__, __blockbusters3.json__ and load the datasets into Pandas DataFrames.


In [3]:
# Transform data in json file into python object
# Blockbusters contains json in row format, while blockbusters2 contains json in columns format
with open("blockbusters.json") as f:
    blockbusters = json.load(f)
with open("blockbusters2.json") as f:
    blockbusters2 = json.load(f)

In [4]:
blockbusters 

[{'title': 'Avengers: Endgame',
  'id': 299534,
  'revenue': 2797800564,
  'genres': [{'id': 12, 'name': 'Adventure'},
   {'id': 878, 'name': 'Science Fiction'},
   {'id': 28, 'name': 'Action'}],
  'belongs_to_collection': {'id': 86311,
   'name': 'The Avengers Collection',
   'poster_path': '/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg',
   'backdrop_path': '/zuW6fOiusv4X9nnW3paHGfXcSll.jpg'},
  'runtime': 181},
 {'title': 'Avatar',
  'id': 19995,
  'revenue': 2787965087,
  'genres': [{'id': 28, 'name': 'Action'},
   {'id': 12, 'name': 'Adventure'},
   {'id': 14, 'name': 'Fantasy'},
   {'id': 878, 'name': 'Science Fiction'}],
  'belongs_to_collection': {'id': 87096,
   'name': 'Avatar Collection',
   'poster_path': '/nslJVsO58Etqkk17oXMuVK4gNOF.jpg',
   'backdrop_path': '/8nCr9W7sKus2q9PLbYsnT7iCkuT.jpg'},
  'runtime': 162},
 {'title': 'Star Wars: The Force Awakens',
  'id': 140607,
  'revenue': 2068223624,
  'genres': [{'id': 28, 'name': 'Action'},
   {'id': 12, 'name': 'Adventure'},
   {'id': 8

In [5]:
blockbusters2

{'title': {'0': 'Avengers: Endgame',
  '1': 'Avatar',
  '2': 'Star Wars: The Force Awakens',
  '3': 'Avengers: Infinity War',
  '4': 'Titanic',
  '5': 'Jurassic World',
  '6': 'The Lion King',
  '7': 'The Avengers',
  '8': 'Furious 7',
  '9': 'Avengers: Age of Ultron',
  '10': 'Black Panther',
  '11': 'Harry Potter and the Deathly Hallows: Part 2',
  '12': 'Star Wars: The Last Jedi',
  '13': 'Frozen II',
  '14': 'Jurassic World: Fallen Kingdom',
  '15': 'Frozen',
  '16': 'Beauty and the Beast',
  '17': 'Incredibles 2'},
 'id': {'0': 299534,
  '1': 19995,
  '2': 140607,
  '3': 299536,
  '4': 597,
  '5': 135397,
  '6': 420818,
  '7': 24428,
  '8': 168259,
  '9': 99861,
  '10': 284054,
  '11': 12445,
  '12': 181808,
  '13': 330457,
  '14': 351286,
  '15': 109445,
  '16': 321612,
  '17': 260513},
 'revenue': {'0': 2797800564,
  '1': 2787965087,
  '2': 2068223624,
  '3': 2046239637,
  '4': 1845034188,
  '5': 1671713208,
  '6': 1656943394,
  '7': 1519557910,
  '8': 1506249360,
  '9': 1405403

In [8]:
# Convert python object blockbusters into pandas dataframes 
blockbusters_df = pd.DataFrame(blockbusters)
blockbusters_df2 = pd.DataFrame(blockbusters2)

# The loading and conversion of json objects into pandas dataframes can also done in one step instead using pd.readjson. This function is also better when converting json that's in a split format into a dataframe.
blockbusters_df3 = pd.read_json("blockbusters3.json", orient="split")

In [9]:
blockbusters_df

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

In [10]:
blockbusters_df2

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

In [11]:
blockbusters_df3

title      id     revenue  \
0                              Avengers: Endgame  299534  2797800564   
1                                         Avatar   19995  2787965087   
2                   Star Wars: The Force Awakens  140607  2068223624   
3                         Avengers: Infinity War  299536  2046239637   
4                                        Titanic     597  1845034188   
5                                 Jurassic World  135397  1671713208   
6                                  The Lion King  420818  1656943394   
7                                   The Avengers   24428  1519557910   
8                                      Furious 7  168259  1506249360   
9                        Avengers: Age of Ultron   99861  1405403694   
10                                 Black Panther  284054  1346739107   
11  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219   
12                      Star Wars: The Last Jedi  181808  1332539889   
13                                     Frozen II  330457  1330764959   
14                Jurassic World: Fallen Kingdom  351286  1303459585   
15                                        Frozen  109445  1274219009   
16                          Beauty and the Beast  321612  1263521126   
17                                 Incredibles 2  260513  1241891456   

                                               genres  \
0   [{'id': 12, 'name': 'Adventure'}, {'id': 878, ...   
1   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
2   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
3   [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   
4   [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
5   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
6   [{'id': 12, 'name': 'Adventure'}, {'id': 10751...   
7   [{'id': 878, 'name': 'Science Fiction'}, {'id'...   
8   [{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...   
9   [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
10  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
11  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
12  [{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...   
13  [{'id': 16, 'name': 'Animation'}, {'id': 10751...   
14  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   
15  [{'id': 16, 'name': 'Animation'}, {'id': 12, '...   
16  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
17  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                belongs_to_collection  runtime  
0   {'id': 86311, 'name': 'The Avengers Collection...      181  
1   {'id': 87096, 'name': 'Avatar Collection', 'po...      162  
2   {'id': 10, 'name': 'Star Wars Collection', 'po...      136  
3   {'id': 86311, 'name': 'The Avengers Collection...      149  
4                                                None      194  
5   {'id': 328, 'name': 'Jurassic Park Collection'...      124  
6                                                None      118  
7   {'id': 86311, 'name': 'The Avengers Collection...      143  
8   {'id': 9485, 'name': 'The Fast and the Furious...      137  
9   {'id': 86311, 'name': 'The Avengers Collection...      141  
10  {'id': 529892, 'name': 'Black Panther Collecti...      134  
11  {'id': 1241, 'name': 'Harry Potter Collection'...      130  
12  {'id': 10, 'name': 'Star Wars Collection', 'po...      152  
13  {'id': 386382, 'name': 'Frozen Collection', 'p...      104  
14  {'id': 328, 'name': 'Jurassic Park Collection'...      129  
15  {'id': 386382, 'name': 'Frozen Collection', 'p...      102  
16                                               None      129  
17  {'id': 468222, 'name': 'The Incredibles Collec...      118

In [31]:
# The data in the blockbusters_df dataframe is nested. Normalise the data. The data in the blockbusters_df2 and blockbusters_df3 dataframes are also nested but cannot be normalised using pd.json_normalize as these dataframes were imported from json files in column and split formats. 

# Normalise belongs_to_collection column
collection_norm_df = pd.json_normalize(data = blockbusters, sep= "_")

# Normalise genres column
genre_norm_df = pd.json_normalize(data = blockbusters, record_path="genres", meta=["id"], record_prefix="genre_")

# Append normalised genre column to dataframe with normalised collection column
blockbusters_norm_df = collection_norm_df.drop(columns="genres").merge(genre_norm_df, how = "inner", on="id")
blockbusters_norm_df

title      id     revenue  runtime  \
0                              Avengers: Endgame  299534  2797800564      181   
1                              Avengers: Endgame  299534  2797800564      181   
2                              Avengers: Endgame  299534  2797800564      181   
3                                         Avatar   19995  2787965087      162   
4                                         Avatar   19995  2787965087      162   
5                                         Avatar   19995  2787965087      162   
6                                         Avatar   19995  2787965087      162   
7                   Star Wars: The Force Awakens  140607  2068223624      136   
8                   Star Wars: The Force Awakens  140607  2068223624      136   
9                   Star Wars: The Force Awakens  140607  2068223624      136   
10                  Star Wars: The Force Awakens  140607  2068223624      136   
11                        Avengers: Infinity War  299536  2046239637      149   
12                        Avengers: Infinity War  299536  2046239637      149   
13                        Avengers: Infinity War  299536  2046239637      149   
14                                       Titanic     597  1845034188      194   
15                                       Titanic     597  1845034188      194   
16                                       Titanic     597  1845034188      194   
17                                Jurassic World  135397  1671713208      124   
18                                Jurassic World  135397  1671713208      124   
19                                Jurassic World  135397  1671713208      124   
20                                Jurassic World  135397  1671713208      124   
21                                 The Lion King  420818  1656943394      118   
22                                 The Lion King  420818  1656943394      118   
23                                  The Avengers   24428  1519557910      143   
24                                  The Avengers   24428  1519557910      143   
25                                  The Avengers   24428  1519557910      143   
26                                     Furious 7  168259  1506249360      137   
27                                     Furious 7  168259  1506249360      137   
28                       Avengers: Age of Ultron   99861  1405403694      141   
29                       Avengers: Age of Ultron   99861  1405403694      141   
30                       Avengers: Age of Ultron   99861  1405403694      141   
31                                 Black Panther  284054  1346739107      134   
32                                 Black Panther  284054  1346739107      134   
33                                 Black Panther  284054  1346739107      134   
34                                 Black Panther  284054  1346739107      134   
35  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219      130   
36  Harry Potter and the Deathly Hallows: Part 2   12445  1341511219      130   
37                      Star Wars: The Last Jedi  181808  1332539889      152   
38                      Star Wars: The Last Jedi  181808  1332539889      152   
39                      Star Wars: The Last Jedi  181808  1332539889      152   
40                      Star Wars: The Last Jedi  181808  1332539889      152   
41                                     Frozen II  330457  1330764959      104   
42                                     Frozen II  330457  1330764959      104   
43                                     Frozen II  330457  1330764959      104   
44                Jurassic World: Fallen Kingdom  351286  1303459585      129   
45                Jurassic World: Fallen Kingdom  351286  1303459585      129   
46                Jurassic World: Fallen Kingdom  351286  1303459585      129   
47                                        Frozen  109445  1274219009      102   
48                                        Frozen  109445  1274219009      

## Working with APIs and JSON (Part 1)

__API-Request__: Load all available information for the movie with __movie id = 140607__ using the TMDB database API and save as a Pandas DataFrame. 

In [7]:
# Generate the URL for an API request for the movie with id = 140607 from the TMDB database
movie_id = "140607"
api_key = "96944f8e497f77db9d20db6e35c77804"
url = "https://api.themoviedb.org/3/movie/" + movie_id + "?api_key=" + api_key + "&language=en-US"
url

'https://api.themoviedb.org/3/movie/140607?api_key=96944f8e497f77db9d20db6e35c77804&language=en-US'

In [8]:
# Request the movie information
request = requests.get(url)
request

<Response [200]>

In [9]:
# Return the request information as a json object
request_json = request.json()
request_json

{'adult': False,
 'backdrop_path': '/8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg',
 'belongs_to_collection': {'id': 10,
  'name': 'Star Wars Collection',
  'poster_path': '/gq5Wi7i4SF3lo4HHkJasDV95xI9.jpg',
  'backdrop_path': '/d8duYyyC9J5T825Hg7grmaabfxQ.jpg'},
 'budget': 245000000,
 'genres': [{'id': 12, 'name': 'Adventure'},
  {'id': 28, 'name': 'Action'},
  {'id': 878, 'name': 'Science Fiction'},
  {'id': 14, 'name': 'Fantasy'}],
 'homepage': 'http://www.starwars.com/films/star-wars-episode-vii',
 'id': 140607,
 'imdb_id': 'tt2488496',
 'original_language': 'en',
 'original_title': 'Star Wars: The Force Awakens',
 'overview': 'Thirty years after defeating the Galactic Empire, Han Solo and his allies face a new threat from the evil Kylo Ren and his army of Stormtroopers.',
 'popularity': 66.464,
 'poster_path': '/wqnLdwVXoBjKibFRR5U3y0aDUhs.jpg',
 'production_companies': [{'id': 1,
   'logo_path': '/o86DbpburjxrqAzEDhXZcyE8pDb.png',
   'name': 'Lucasfilm Ltd.',
   'origin_country': 'US'},
  {'i

In [17]:
# Convert the json object into a pandas dataframe
starwars_df = pd.Series(request_json).to_frame().transpose()
starwars_df

adult                     backdrop_path  \
0  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg   

                               belongs_to_collection     budget  \
0  {'id': 10, 'name': 'Star Wars Collection', 'po...  245000000   

                                              genres  \
0  [{'id': 12, 'name': 'Adventure'}, {'id': 28, '...   

                                            homepage      id    imdb_id  \
0  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   

  original_language                original_title  ... release_date  \
0                en  Star Wars: The Force Awakens  ...   2015-12-15   

      revenue runtime                                   spoken_languages  \
0  2068223624     136  [{'english_name': 'English', 'iso_639_1': 'en'...   

     status                        tagline                         title  \
0  Released  Every generation has a story.  Star Wars: The Force Awakens   

   video vote_average vote_count  
0  False         7.31      17472  

[1 rows x 25 columns]

In [42]:
# Normalise data
# Normalise columns separately
belongs_to_collection_norm = pd.json_normalize(request_json).drop(columns="genres")
genres_norm = pd.json_normalize(request_json,record_path="genres",meta=["title"],record_prefix="genre_")
production_companies_norm = pd.json_normalize(request_json,record_path="production_companies",meta="title",record_prefix="production_company_")
production_countries_norm = pd.json_normalize(request_json,record_path="production_countries",meta="title",record_prefix="production_country_")
spoken_languages_norm = pd.json_normalize(request_json,record_path="spoken_languages",meta="title",record_prefix="spoken_language_")

# Combine normalised data using partial and reduce functions. Partial freezes a portion of a function's arguments and keywords, while reduce allows appliation of an iterable list of objects to the partial object.
dfs = [belongs_to_collection_norm, genres_norm, production_companies_norm, production_countries_norm, spoken_languages_norm]
merge = partial(pd.merge, on="title")
starwars_df_norm = reduce(merge,dfs)
starwars_df_norm


adult                     backdrop_path     budget  \
0  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
1  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
2  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
3  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
4  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
5  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
6  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   
7  False  /8BTsTfln4jlQrLXUBquXJ0ASQy9.jpg  245000000   

                                            homepage      id    imdb_id  \
0  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
1  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
2  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
3  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
4  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
5  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
6  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   
7  http://www.starwars.com/films/star-wars-episod...  140607  tt2488496   

  original_language                original_title  \
0                en  Star Wars: The Force Awakens   
1                en  Star Wars: The Force Awakens   
2                en  Star Wars: The Force Awakens   
3                en  Star Wars: The Force Awakens   
4                en  Star Wars: The Force Awakens   
5                en  Star Wars: The Force Awakens   
6                en  Star Wars: The Force Awakens   
7                en  Star Wars: The Force Awakens   

                                            overview  popularity  ...  \
0  Thirty years after defeating the Galactic Empi...      66.464  ...   
1  Thirty years after defeating the Galactic Empi...      66.464  ...   
2  Thirty years after defeating the Galactic Empi...      66.464  ...   
3  Thirty years after defeating the Galactic Empi...      66.464  ...   
4  Thirty years after defeating the Galactic Empi...      66.464  ...   
5  Thirty years after defeating the Galactic Empi...      66.464  ...   
6  Thirty years after defeating the Galactic Empi...      66.464  ...   
7  Thirty years after defeating the Galactic Empi...      66.464  ...   

        genre_name production_company_id      production_company_logo_path  \
0        Adventure                     1  /o86DbpburjxrqAzEDhXZcyE8pDb.png   
1        Adventure                 11461  /p9FoEt5shEKRWRKVIlvFaEmRnun.png   
2           Action                     1  /o86DbpburjxrqAzEDhXZcyE8pDb.png   
3           Action                 11461  /p9FoEt5shEKRWRKVIlvFaEmRnun.png   
4  Science Fiction                     1  /o86DbpburjxrqAzEDhXZcyE8pDb.png   
5  Science Fiction                 11461  /p9FoEt5shEKRWRKVIlvFaEmRnun.png   
6          Fantasy                     1  /o86DbpburjxrqAzEDhXZcyE8pDb.png   
7          Fantasy                 11461  /p9FoEt5shEKRWRKVIlvFaEmRnun.png   

  production_company_name  production_company_origin_country  \
0          Lucasfilm Ltd.                                 US   
1               Bad Robot                                 US   
2          Lucasfilm Ltd.                                 US   
3               Bad Robot                                 US   
4          Lucasfilm Ltd.                                 US   
5               Bad Robot                                 US   
6          Lucasfilm Ltd.                                 US   
7               Bad Robot                                 US   

   production_country_iso_3166_1   production_country_name  \
0                             US  United States of America   
1                             US  United States of America   
2                             US  United States of America   
3                             US  United States of America   
4                             US  United States of America   
5                        

## Working with APIs and JSON (Part 2)

5. __API-Request__ (discover module): Load all movies with __release date between 2020-01-01 and 2020-02-29__ into a Pandas DataFrame. <br>
See https://www.themoviedb.org/documentation/api/discover and https://developers.themoviedb.org/3/discover/movie-discover for more information.

##  Importing and Saving the Movies Dataset (Best Practice)

6. __API-Request__ (movie module): Load all available information for the movies with movie id = [__299534, 19995, 140607, 299536, 597, 135397, 420818, 24428, 168259, 99861, 284054, 12445, 181808, 330457, 351286, 109445, 321612, 260513__] into a Pandas DataFrame and __save the dataset in a local json file__.

# +++++++++ See some Hints below +++++++++++++

# ++++++++++++++ Hints +++++++++++++++++++++

__Hints for 1.__ <br>
To load json files you can use 

In [ ]:
with open("filename.json") as f:
    data = json.load(f)

and 

In [ ]:
pd.DataFrame(data), pd.read_json(filename.json), pd.json_normalize(data)

the json files have the following orientation (important when using pd.read_json()):
- blockbusters.json -> record
- blockbusters2.json -> column
- blockbusters3.json -> split 

__Hints for 4., 5., 6.__<br>
Make API GET-requests with the library requests (import requests):

In [ ]:
data = requests.get(url).json()

__Hints for 4. and 6.,__ <br> url structure for movie module:

"https://api.themoviedb.org/3/movie/insert_movie_id?api_key=insert_api_key" (replace "insert_movie_id" with movie id and "insert_api_key" with your personal api-key)

__Hints for 5.__<br>
url structure for discover module:

"https://api.themoviedb.org/3/discover/movie?api_key=insert_api_key&query1&query2..." (replace "insert_api_key" with your personal api-key and add appropriate queries)